File: src/main/java/io/github/rcarlosdasilva/cogen/core/handler/TargetFilterHandler.java
package io.github.rcarlosdasilva.cogen.core.handler;

import com.google.common.base.Strings;
import com.google.common.collect.Lists;
import io.github.rcarlosdasilva.cogen.config.convention.JavaType;
import io.github.rcarlosdasilva.cogen.config.various.Database;
import io.github.rcarlosdasilva.cogen.config.various.Table;
import io.github.rcarlosdasilva.cogen.converter.DbTypeConverter;
import io.github.rcarlosdasilva.cogen.core.Handler;
import io.github.rcarlosdasilva.cogen.core.Storage;
import io.github.rcarlosdasilva.cogen.model.FieldDetail;
import io.github.rcarlosdasilva.cogen.model.TableDetail;
import io.github.rcarlosdasilva.cogen.utils.LogIf;
import io.github.rcarlosdasilva.cogen.utils.Utils;
import io.github.rcarlosdasilva.kits.string.TextHelper;
import lombok.extern.slf4j.Slf4j;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * @author <a href="mailto:rcarlosdasilva@qq.com">Dean Zhao</a>
 */
@Slf4j
public class TargetFilterHandler implements Handler {

  private Database database;

  @Override
  public void handle() {
    log.info("[filter] - 暖床ing");

    database = Storage.getConfiguration().getDatabase();

    // 从数据库取回来的
    List<TableDetail> tableDetails = Storage.getAllTables();

    if (database != null) {
      LogIf.error(log, tableDetails == null, "没有获取任何表信息");

      // 配置的表
      Map<String, Table> tableConfigs = database.getTables();
      List<TableDetail> results = Lists.newArrayList();

      tableDetails.forEach(detail -> {
        tableConfigs.forEach((prefix, tableConfig) -> {
          TableDetail td = tableHits(detail, prefix, tableConfig);
          if (td != null) {
            results.add(td);
          }
        });
      });

      Storage.setFilteredTables(results);
    }
  }

  private TableDetail tableHits(TableDetail tableDetail, String prefix, Table table) {
    String name = tableDetail.getName();
    if (Strings.isNullOrEmpty(name) || !name.startsWith(prefix)) {
      return null;
    }

    if (table.getIncluds() != null) {
      if (!table.getIncluds().contains(name)) {
        return null;
      }
    } else if (table.getExcluds() != null) {
      if (table.getExcluds().contains(name)) {
        return null;
      }
    }

    final String cut = table.isHoldTablePrefix() ? "" : table.getPrefix();
    tableDetail.setBaseName(Utils.name(tableDetail.getName(), cut, true));

    List<FieldDetail> fieldDetails = tableDetail.getFields();
    LogIf.error(log, fieldDetails == null, "读取不到表字段信息");
    List<FieldDetail> results = Lists.newArrayList();
    fieldDetails.forEach(detail -> {
      FieldDetail fd = fieldHits(detail, table);
      if (fd != null) {
        results.add(fd);
      }
    });
    tableDetail.setFields(results);

    return tableDetail;
  }

  private FieldDetail fieldHits(FieldDetail fieldDetail, Table table) {
    if (fieldDetail.isPrimaryKey() && database.isIgnoreId()) {
      return null;
    }
    if (database.getIgnoreFields() != null && database.getIgnoreFields().contains(fieldDetail.getName())) {
      return null;
    }
    if (table.getIgnoreFields() != null && table.getIgnoreFields().contains(fieldDetail.getName())) {
      return null;
    }

    if (database.getIgnoreFieldsByPrefix() != null) {
      Optional<String> result = database.getIgnoreFieldsByPrefix().stream().filter(prefix ->
          fieldDetail.getName().startsWith(prefix)).findFirst();
      if (result.isPresent()) {
        return null;
      }
    }

    String baseName = null;
    if (!table.isHoldFieldPrefix()) {
      Optional<String> result = table.getFieldPrefixs().stream().filter(prefix -> fieldDetail.getName().startsWith
          (prefix)).findFirst();
      if (result.isPresent()) {
        baseName = Utils.name(fieldDetail.getName(), result.get(), false);
      }
    }
    if (baseName == null) {
      baseName = Utils.name(fieldDetail.getName(), false);
    }

    fieldDetail.setBaseName(baseName);
    fieldDetail.setJavaType(database.getDbTypeConverter().convert(fieldDetail.getType()));
    return fieldDetail;
  }

}

File: src/main/kotlin/io/github/rcarlosdasilva/cogen/core/handler/TargetFilterHandler.kt
package io.github.rcarlosdasilva.cogen.core.handler

import com.google.common.collect.Lists
import io.github.rcarlosdasilva.cogen.config.Table
import io.github.rcarlosdasilva.cogen.core.Handler
import io.github.rcarlosdasilva.cogen.core.Storage
import io.github.rcarlosdasilva.cogen.model.FieldModel
import io.github.rcarlosdasilva.cogen.model.TableModel
import io.github.rcarlosdasilva.cogen.util.LogIf
import io.github.rcarlosdasilva.cogen.util.name
import mu.KotlinLogging

object TargetFilterHandler : Handler {

  private val logger = KotlinLogging.logger {}

  override fun handle() {
    logger.info("[filter] - 卸妆ing")

    Storage.configuration!!.database?.run {
      LogIf.error(logger, Storage.tables == null, "没有获取任何表信息")
      val results = mutableListOf<TableModel>()

      // 从数据库取回来的
      Storage.tables?.forEach { model ->
        // 配置的表
        tables.forEach { (prefix, tableConfig) ->
          tableHits(model, prefix, tableConfig)?.also {
            results.add(it)
          }
        }
      }

      Storage.filteredTables = results
    }
  }

  private fun tableHits(tableModel: TableModel, prefix: String, tableConfig: Table): TableModel? {
    val name = tableModel.name
    if (name.isBlank() || !name.startsWith(prefix)) {
      return null
    }

    tableConfig.includes?.run {
      if (!this.contains(name)) {
        return null
      }
    }
    tableConfig.excludes?.run {
      if (this.contains(name)) {
        return null
      }
    }

    val cut = if (tableConfig.isHoldTablePrefix) "" else tableConfig.prefix
    tableModel.javaName = name(name, cut, true)

    LogIf.error(logger, tableModel.fields == null, "读取不到表字段信息")
    val results = Lists.newArrayList<FieldModel>()

    tableModel.fields?.forEach { model ->
      fieldHits(model, tableConfig).also {
        results.add(it)
      }
    }
    tableModel.fields = results

    return tableModel
  }
}

private fun fieldHits(fieldModel: FieldModel, table: Table): FieldModel {
  with(Storage.configuration!!.database!!) {
    // 字段是否需要忽略，主键忽略规则1个，全局忽略规则2个，表内规则1个
    val ignore = when {
      fieldModel.isPrimaryKey && isIgnoreId -> true
      ignoreFields != null && ignoreFields!!.contains(fieldModel.name) -> true
      ignoreFieldsByPrefixes?.firstOrNull { fieldModel.name.startsWith(it) } != null -> true
      table.ignoreFields != null && table.ignoreFields!!.contains(fieldModel.name) -> true
      else -> false
    }

    if (ignore) {
      fieldModel.isIgnore = ignore
      return fieldModel
    }

    val cut = cutFieldPrefixes?.firstOrNull {
      fieldModel.name.startsWith(it)
    } ?: table.cutFieldPrefixes?.firstOrNull {
      fieldModel.name.startsWith(it)
    }
    fieldModel.javaName = cut?.let {
      name(fieldModel.name, it, false)
    } ?: name(fieldModel.name, false)

    fieldModel.javaType = dbTypeConverter.convert(fieldModel.type!!)
  }
  return fieldModel
}